Source/Core/VideoCommon/VertexManagerBase.cpp

```diff
@@ -610,7 +640,65 @@ void VertexManagerBase::Flush()
     UpdatePipelineObject();
     if (m_current_pipeline_object)
     {
-      g_gfx->SetPipeline(m_current_pipeline_object);
+      const AbstractPipeline* current_pipeline = m_current_pipeline_object;
+      if (!custom_pixel_shader_contents.shaders.empty())
+      {
+        CustomShaderInstance custom_shaders;
+        custom_shaders.pixel_contents = std::move(custom_pixel_shader_contents);
+
+        switch (g_ActiveConfig.iShaderCompilationMode)
+        {
+        case ShaderCompilationMode::Synchronous:
+        case ShaderCompilationMode::AsynchronousSkipRendering:
+        {
+          if (auto pipeline = m_custom_shader_cache->GetPipelineAsync(
+                  m_current_pipeline_config, custom_shaders, m_current_pipeline_object->m_config))
+          {
+            current_pipeline = *pipeline;
+          }
+        }
+        break;
+        case ShaderCompilationMode::SynchronousUberShaders:
+        {
+          // D3D has issues compiling large custom ubershaders
+          // use synchronous instead
```

Comment:
I believe the best term here would be "specialized shaders" instead - that's the one we use on the UI, though the enum hasn't been changed for it.

Also, I'm not entirely sure why all of these use `GetPipelineAsync`. What happens if the pipeline isn't present? Does it render without the custom shaders?

Reply:
> Also, I'm not entirely sure why all of these use GetPipelineAsync. What happens if the pipeline isn't present? Does it render without the custom shaders?

Yes, if there is an error, we fall back to the game shader